Kudu.Core/Functions/FunctionManager.cs

```diff
@@ -191,34 +199,71 @@ internal static JArray GetTriggers(IEnumerable<FunctionEnvelope> functions, ITra
             return config;
         }
 
-        public async Task<FunctionSecrets> GetFunctionSecretsAsync(string functionName)
+        private async Task<T> KeyOpHelper<T>(string name, IKeyJsonOps<T> keyOp)
         {
-            FunctionSecrets secrets;
-            string secretFilePath = GetFunctionSecretsFilePath(functionName);
-            if (FileSystemHelpers.FileExists(secretFilePath))
+            string keyPath = GetFunctionSecretsFilePath(name);
+            string key = null;
+            if (!FileSystemHelpers.FileExists(keyPath))
             {
-                // load the secrets file
-                string secretsJson = await FileSystemHelpers.ReadAllTextFromFileAsync(secretFilePath);
-                secrets = JsonConvert.DeserializeObject<FunctionSecrets>(secretsJson);
+                FileSystemHelpers.EnsureDirectory(Path.GetDirectoryName(keyPath));
+                try
+                {
+                    using (var fileStream = FileSystemHelpers.OpenFile(keyPath, FileMode.CreateNew, FileAccess.Write, FileShare.None))
+                    // will fail if file exists, prevent reading prematurely
+                    // getting the lock early so no redundant work is being done
+                    {
+                        string jsonContent = keyOp.GenerateKeyUglyJson(SecurityUtility.GenerateSecretStringsKeyPair(keyOp.RequireKeyCount()), FunctionSiteExtensionVersion, out key);
+                        using (var sw = new StringWriter())
+                        using (var sr = new System.IO.StringReader(jsonContent))
+                        {
+                            new JsonTextWriter(sw) { Formatting = Formatting.Indented }.WriteToken(new JsonTextReader(sr));
+                            // if lock acquire lock return false, I wait until write finishes and read keyPath
+                            using (var streamWriter = new StreamWriter(fileStream))
+                            {
+                                await streamWriter.WriteAsync(sw.ToString());
+                                await streamWriter.FlushAsync();
+                            }
+                        }
+                    }
+                    return keyOp.GenerateKeyObject(key, name);
+                }
+                catch (Exception)
+                {
+                    // fallback to read key files
+                }
             }
-            else
+
+            for (int timeOut = 5; timeOut >= 0; timeOut--)
```

Comment:
The retry loop should only be around the reading of the file. This way, if other random things blow up (e.g. DecryptSecretString), it won't look like we're dealing with a file being held issue.